File: files/detection/tracker.py
"""
Multi-Object Tracker Module

Implements DeepSORT-style tracking for consistent player and ball IDs
across frames. Maintains track continuity even during occlusions.
"""

import numpy as np
from typing import List, Dict, Optional
from scipy.optimize import linear_sum_assignment
from filterpy.kalman import KalmanFilter


class Track:
    """
    Represents a tracked object across multiple frames.
    """
    
    next_id = 0
    
    def __init__(self, bbox: List[float], class_type: str, confidence: float):
        """
        Initialize track.
        
        Args:
            bbox: [x1, y1, x2, y2] bounding box
            class_type: Object class (player/ball/racket)
            confidence: Detection confidence
        """
        self.track_id = Track.next_id
        Track.next_id += 1
        
        self.class_type = class_type
        self.bbox = bbox
        self.confidence = confidence
        
        # Tracking state
        self.age = 0  # Frames since creation
        self.hits = 1  # Number of detections
        self.time_since_update = 0
        
        # Kalman filter for smooth trajectory
        self.kf = self._init_kalman_filter(bbox)
        
        # Appearance feature (for re-identification)
        self.appearance_features = []
    
    def _init_kalman_filter(self, bbox: List[float]) -> KalmanFilter:
        """Initialize Kalman filter for bbox tracking."""
        kf = KalmanFilter(dim_x=7, dim_z=4)
        
        # State: [x, y, w, h, vx, vy, vw] (position, size, velocities)
        # Measurement: [x, y, w, h]
        
        x, y = (bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2
        w, h = bbox[2] - bbox[0], bbox[3] - bbox[1]
        
        kf.x = np.array([x, y, w, h, 0, 0, 0])
        
        # Measurement matrix
        kf.H = np.array([
            [1, 0, 0, 0, 0, 0, 0],
            [0, 1, 0, 0, 0, 0, 0],
            [0, 0, 1, 0, 0, 0, 0],
            [0, 0, 0, 1, 0, 0, 0]
        ])
        
        # Measurement noise
        kf.R *= 10
        
        # Process noise
        kf.Q[-1, -1] *= 0.01
        kf.Q[4:, 4:] *= 0.01
        
        # Covariance
        kf.P[4:, 4:] *= 1000
        kf.P *= 10
        
        return kf
    
    def predict(self):
        """Predict next state using Kalman filter."""
        self.kf.predict()
        self.age += 1
        self.time_since_update += 1
    
    def update(self, bbox: List[float], confidence: float):
        """
        Update track with new detection.
        
        Args:
            bbox: New bounding box
            confidence: Detection confidence
        """
        x, y = (bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2
        w, h = bbox[2] - bbox[0], bbox[3] - bbox[1]
        
        self.kf.update(np.array([x, y, w, h]))
        
        self.bbox = bbox
        self.confidence = confidence
        self.hits += 1
        self.time_since_update = 0
    
    def get_state(self) -> Dict:
        """Get current track state."""
        x, y, w, h = self.kf.x[:4]
        
        bbox = [
            x - w/2,  # x1
            y - h/2,  # y1
            x + w/2,  # x2
            y + h/2   # y2
        ]
        
        return {
            'track_id': self.track_id,
            'bbox': bbox,
            'class': self.class_type,
            'confidence': self.confidence,
            'age': self.age,
            'hits': self.hits
        }


class MultiObjectTracker:
    """
    DeepSORT-style multi-object tracker.
    
    Maintains consistent IDs for players and ball across frames,
    handling occlusions and re-appearances.
    """
    
    def __init__(self, config: Dict):
        """
        Initialize tracker.
        
        Args:
            config: Tracking configuration
        """
        self.max_age = config.get('max_age', 30)
        self.min_hits = config.get('min_hits', 3)
        self.iou_threshold = config.get('iou_threshold', 0.3)
        self.use_appearance = config.get('use_appearance', False)
        
        self.tracks = []
    
    def update(
        self,
        detections: Dict[str, List],
        frame: Optional[np.ndarray] = None
    ) -> List[Dict]:
        """
        Update tracks with new detections.
        
        Args:
            detections: Detection dictionary from detector
            frame: Optional frame for appearance features
        
        Returns:
            List of active track states
        """
        # Predict new locations for existing tracks
        for track in self.tracks:
            track.predict()
        
        # Convert detections to list
        det_list = []
        for det_type in ['players', 'ball', 'rackets']:
            for det in detections.get(det_type, []):
                det_list.append({
                    'bbox': det['bbox'],
                    'confidence': det['confidence'],
                    'class': det_type[:-1] if det_type.endswith('s') else det_type
                })
        
        # Associate detections to tracks
        matches, unmatched_dets, unmatched_tracks = self._associate(det_list)
        
        # Update matched tracks
        for track_idx, det_idx in matches:
            det = det_list[det_idx]
            self.tracks[track_idx].update(det['bbox'], det['confidence'])
        
        # Create new tracks for unmatched detections
        for det_idx in unmatched_dets:
            det = det_list[det_idx]
            new_track = Track(det['bbox'], det['class'], det['confidence'])
            self.tracks.append(new_track)
        
        # Remove old tracks
        self.tracks = [
            t for t in self.tracks
            if t.time_since_update <= self.max_age
        ]
        
        # Return confirmed tracks
        active_tracks = [
            t.get_state()
            for t in self.tracks
            if t.hits >= self.min_hits or t.age < self.min_hits
        ]
        
        return active_tracks
    
    def _associate(
        self,
        detections: List[Dict]
    ) -> tuple:
        """
        Associate detections to tracks using IoU and appearance.
        
        Args:
            detections: List of detection dicts
        
        Returns:
            Tuple of (matches, unmatched_detections, unmatched_tracks)
        """
        if len(self.tracks) == 0:
            return [], list(range(len(detections))), []
        
        if len(detections) == 0:
            return [], [], list(range(len(self.tracks)))
        
        # Compute IoU cost matrix
        iou_matrix = np.zeros((len(self.tracks), len(detections)))
        
        for t_idx, track in enumerate(self.tracks):
            track_bbox = track.get_state()['bbox']
            for d_idx, det in enumerate(detections):
                # Only match same class
                if track.class_type == det['class']:
                    iou = self._compute_iou(track_bbox, det['bbox'])
                    iou_matrix[t_idx, d_idx] = iou
        
        # Hungarian algorithm for optimal assignment
        # Convert IoU to cost (1 - IoU)
        cost_matrix = 1 - iou_matrix
        
        track_indices, det_indices = linear_sum_assignment(cost_matrix)
        
        # Filter out low IoU matches
        matches = []
        for t_idx, d_idx in zip(track_indices, det_indices):
            if iou_matrix[t_idx, d_idx] >= self.iou_threshold:
                matches.append((t_idx, d_idx))
        
        # Find unmatched detections and tracks
        matched_track_indices = [m[0] for m in matches]
        matched_det_indices = [m[1] for m in matches]
        
        unmatched_tracks = [
            i for i in range(len(self.tracks))
            if i not in matched_track_indices
        ]
        
        unmatched_dets = [
            i for i in range(len(detections))
            if i not in matched_det_indices
        ]
        
        return matches, unmatched_dets, unmatched_tracks
    
    @staticmethod
    def _compute_iou(bbox1: List[float], bbox2: List[float]) -> float:
        """
        Compute Intersection over Union of two bboxes.
        
        Args:
            bbox1: [x1, y1, x2, y2]
            bbox2: [x1, y1, x2, y2]
        
        Returns:
            IoU value [0, 1]
        """
        x1_tl = max(bbox1[0], bbox2[0])
        y1_tl = max(bbox1[1], bbox2[1])
        x1_br = min(bbox1[2], bbox2[2])
        y1_br = min(bbox1[3], bbox2[3])
        
        if x1_br < x1_tl or y1_br < y1_tl:
            return 0.0
        
        intersection = (x1_br - x1_tl) * (y1_br - y1_tl)
        
        area1 = (bbox1[2] - bbox1[0]) * (bbox1[3] - bbox1[1])
        area2 = (bbox2[2] - bbox2[0]) * (bbox2[3] - bbox2[1])
        
        union = area1 + area2 - intersection
        
        if union == 0:
            return 0.0
        
        return intersection / union
    
    def reset(self):
        """Reset tracker state."""
        self.tracks = []
        Track.next_id = 0

File: internal/tracker/tracker.go
package tracker

import (
	"github.com/liamp/tennis-tagger/internal/bridge"
)

// TrackedObject represents a tracked object with its current state.
type TrackedObject struct {
	ID              int
	BBox            bridge.BBox
	HitCount        int
	TimeSinceUpdate int
	Age             int
}

// track is the internal representation of a tracked object.
type track struct {
	id              int
	kf              *KalmanFilter
	hitCount        int
	timeSinceUpdate int
	age             int
	lastConfidence  float64
}

// MultiObjectTracker tracks multiple objects across frames using Kalman filters and IoU assignment.
type MultiObjectTracker struct {
	maxAge       int
	minHits      int
	iouThreshold float64
	tracks       []*track
	nextID       int
}

// NewTracker creates a new MultiObjectTracker with the given parameters.
// maxAge: number of frames to keep a track alive without detections.
// minHits: minimum number of hits before a track is returned as active.
// iouThreshold: minimum IoU for a detection to be matched to a track.
func NewTracker(maxAge, minHits int, iouThreshold float64) *MultiObjectTracker {
	return &MultiObjectTracker{
		maxAge:       maxAge,
		minHits:      minHits,
		iouThreshold: iouThreshold,
		nextID:       1,
	}
}

// bboxToXYWH converts a bridge.BBox [x1,y1,x2,y2] to [cx, cy, w, h].
func bboxToXYWH(b bridge.BBox) [4]float64 {
	cx := (b.X1 + b.X2) / 2
	cy := (b.Y1 + b.Y2) / 2
	w := b.X2 - b.X1
	h := b.Y2 - b.Y1
	return [4]float64{cx, cy, w, h}
}

// xywhToX1Y1X2Y2 converts [cx, cy, w, h] to [x1, y1, x2, y2].
func xywhToX1Y1X2Y2(xywh [4]float64) [4]float64 {
	return [4]float64{
		xywh[0] - xywh[2]/2,
		xywh[1] - xywh[3]/2,
		xywh[0] + xywh[2]/2,
		xywh[1] + xywh[3]/2,
	}
}

// Update processes a new set of detections and returns the list of active tracked objects.
func (t *MultiObjectTracker) Update(detections []bridge.BBox) []TrackedObject {
	// 1. Predict all existing tracks
	for _, tr := range t.tracks {
		tr.kf.Predict()
	}

	// 2. Build bbox arrays in [x1,y1,x2,y2] format for assignment
	trackBoxes := make([][4]float64, len(t.tracks))
	for i, tr := range t.tracks {
		kfBBox := tr.kf.BBox() // [cx, cy, w, h]
		trackBoxes[i] = xywhToX1Y1X2Y2(kfBBox)
	}

	detBoxes := make([][4]float64, len(detections))
	for i, d := range detections {
		detBoxes[i] = [4]float64{d.X1, d.Y1, d.X2, d.Y2}
	}

	// 3. Assign detections to tracks
	matches, unmatchedDets, _ := Assign(trackBoxes, detBoxes, t.iouThreshold)

	// 4. Update matched tracks
	for _, m := range matches {
		tr := t.tracks[m.TrackIdx]
		det := detections[m.DetectionIdx]
		measurement := bboxToXYWH(det)
		tr.kf.Update(measurement)
		tr.hitCount++
		tr.timeSinceUpdate = 0
		tr.age++
		tr.lastConfidence = det.Confidence
	}

	// 5. Mark unmatched tracks
	matchedTrackSet := make(map[int]bool)
	for _, m := range matches {
		matchedTrackSet[m.TrackIdx] = true
	}
	for i, tr := range t.tracks {
		if !matchedTrackSet[i] {
			tr.timeSinceUpdate++
			tr.age++
		}
	}

	// 6. Create new tracks for unmatched detections
	for _, di := range unmatchedDets {
		det := detections[di]
		measurement := bboxToXYWH(det)
		kf := NewKalmanFilter(measurement)
		tr := &track{
			id:              t.nextID,
			kf:              kf,
			hitCount:        1,
			timeSinceUpdate: 0,
			age:             1,
			lastConfidence:  det.Confidence,
		}
		t.nextID++
		t.tracks = append(t.tracks, tr)
	}

	// 7. Remove dead tracks (timeSinceUpdate > maxAge)
	alive := make([]*track, 0, len(t.tracks))
	for _, tr := range t.tracks {
		if tr.timeSinceUpdate <= t.maxAge {
			alive = append(alive, tr)
		}
	}
	t.tracks = alive

	// 8. Return active tracks (hitCount >= minHits)
	var result []TrackedObject
	for _, tr := range t.tracks {
		if tr.hitCount >= t.minHits {
			kfBBox := tr.kf.BBox() // [cx, cy, w, h]
			xyxy := xywhToX1Y1X2Y2(kfBBox)
			result = append(result, TrackedObject{
				ID: tr.id,
				BBox: bridge.BBox{
					X1:         xyxy[0],
					Y1:         xyxy[1],
					X2:         xyxy[2],
					Y2:         xyxy[3],
					Confidence: tr.lastConfidence,
				},
				HitCount:        tr.hitCount,
				TimeSinceUpdate: tr.timeSinceUpdate,
				Age:             tr.age,
			})
		}
	}

	return result
}

// Reset clears all tracks and resets the ID counter.
func (t *MultiObjectTracker) Reset() {
	t.tracks = nil
	t.nextID = 1
}
